### ISSUE TITLE
or_fun_call doesn't trigger on unsafe blocks

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#6675]
I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ece94c8fe5548da6f17cf014779fce6)):

```rust
unsafe fn f() {
    None.unwrap_or(String::new().as_mut_vec());
}

fn g() {
    None.unwrap_or(unsafe { String::new().as_mut_vec() });
}
```

I expected to see this happen: Clippy warns on both `f` and `g`.

Instead, this happened: Clippy only warns on `f`.

### Meta

- `cargo clippy -V`: 0.1.51 (2021-02-03 e708cbd)

**Comment by jyn514** [comment]
Also the suggestion causes a borrowck error:
```
error[E0500]: closure requires unique access to `this` but it is already borrowed
    --> src/simple_api/mod.rs:1150:55
     |
1147 |         unsafe fn get_last_buffer(this: &mut Key) -> &mut Vec<u8> {
     |                                         - let's call the lifetime of this reference `'1`
...
1150 |             this.subscripts.last_mut().unwrap_or_else(|| this.variable.as_mut_vec())
     |             ------------------------------------------^^----------------------------
     |             |                                         |  |
     |             |                                         |  second borrow occurs due to use of `this` in closure
     |             |                                         closure construction occurs here
     |             borrow occurs here
     |             returning this value requires that `this.subscripts` is borrowed for `'1`
```

**Comment by camsteffen** [comment]
@jyn514 Can I ask a favor? In the future, it would be nice to have multiple issues reported in separate github issues, even if they are related. It makes it easier to track as they can be fixed separately. Also each bug may have a different difficulty level (which I believe is the case here) and it's nice to have "good first issues".

**Comment by jyn514** [comment]
Sure thing, I can open a new issue.
